File: node/src/main/java/dev.httpmarco.polocloud.node/commands/specific/ServiceArgument.java
package dev.httpmarco.polocloud.node.commands.specific;

import dev.httpmarco.polocloud.api.services.ClusterService;
import dev.httpmarco.polocloud.node.Node;
import dev.httpmarco.polocloud.node.commands.CommandArgument;
import dev.httpmarco.polocloud.node.commands.CommandContext;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public final class ServiceArgument extends CommandArgument<ClusterService> {

    public ServiceArgument(String key) {
        super(key);
    }

    @Override
    public boolean predication(@NotNull String rawInput) {
        return Node.instance().serviceProvider().services().stream().anyMatch(it -> it.name().equalsIgnoreCase(rawInput));
    }

    @Override
    public List<String> defaultArgs(CommandContext context) {
        return Node.instance().serviceProvider().services().stream().map(ClusterService::name).toList();
    }

    @Override
    public ClusterService buildResult(String input) {
        return Node.instance().serviceProvider().find(input);
    }
}

File: agent/src/main/java/dev/httpmarco/polocloud/agent/runtime/local/terminal/commands/type/ServiceArgument.kt
package dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.type

import dev.httpmarco.polocloud.agent.Agent
import dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.CommandArgument
import dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.CommandContext
import dev.httpmarco.polocloud.agent.services.Service

class ServiceArgument(key: String = "service") : CommandArgument<Service>(key) {

    override fun buildResult(input: String, context: CommandContext): Service {
        // null check is done in the predication method
        return Agent.instance.runtime.serviceStorage().findService(input)!!
    }

    override fun defaultArgs(context: CommandContext): MutableList<String> {
        return Agent.instance.runtime.serviceStorage().items().stream().map { it.name() }.toList()
    }

    override fun predication(rawInput: String): Boolean {
        return Agent.instance.runtime.serviceStorage().findService(rawInput) != null
    }
}